### DISCUSSION TITLE
Undo Implementation

### DISCUSSION THREAD

**Discussion by kgodey** [mathesar-foundation/mathesar#113] [Maintainer discussions]
We'd like to implement some form of "undo" so that the user can get back to previous states and recover from errors.

We need to define what actions we'd like to be able to undo before we can finalize implementation details, but I wanted to start a discussion so that we'd start thinking about it.

**Comment by kgodey** [comment]
_Some notes that @mathemancer wrote up when we initially conceived Mathesar._

---

We should keep an audit table for each table/view that's modifiable in a user's database. Two options are possible for this implementation:
1. Use triggers to record logs of each modification at the DB level.
	- See https://wiki.postgresql.org/wiki/Audit_trigger_91plus
2. At the python (service) level, record each operation on the DB and its success on another table.

The advantage of (1) is that we get transactions and consistency guarantees from the DB system. However, this may involve a fair bit of logic, and it won't be tested if it's in the DB rather than python.

For the Audit table itself, our current favored options are:
1. A global audit table with one row per record that's ever existed in a user's database (or schema). The row would have an identifier for the record, and metadata about its history (and location in the DB):
	- last modified
	- created
	- modification history
2. A per-table audit table with the same info as above; easier to find relevant records, and probably a faster join with the relevant table. Downside is we can't track records split across multiple tables, or moved from table A to table B.
3. A global audit table, and a per-table audit (materialized) view for performance.

**Comment by mathemancer** [comment]
I think it's likely (inevitable, probably) that the best way to do implement "undo" (and auditing) will be different for DDL (data definition language) and DML (data manipulation language).
